### PR TITLE
Advanced http filters

### LINKED ISSUE #2683: Support specifying both header and path filter for stealing HTTP
IMO we should probably just allow for any set/combination of filters.
Remember to verify config after fetching `mirrord-protocol` version used by agent/operator

### PR DESCRIPTION
**Pull request by gememma**
Closes #2683

### CODE REVIEW

File: mirrord/agent/src/steal/http/filter.rs [hunk 1]
```diff
@@ -22,6 +29,17 @@ impl TryFrom<&mirrord_protocol::tcp::HttpFilter> for HttpFilter {
             mirrord_protocol::tcp::HttpFilter::Path(path) => {
                 Ok(Self::Path(Regex::new(&format!("(?i){path}"))?))
             }
+            mirrord_protocol::tcp::HttpFilter::Composite { all, filters } => {
+                let all = *all;
+                let mut agent_filters = vec![];
+                for filter in filters {
+                    agent_filters.push(TryFrom::try_from(filter)?);
+                }
```
Comment (Razz4780) [inline mirrord/agent/src/steal/http/filter.rs:37]:
```suggestion
                let filters = filters.into_iter().map(HttpFilter::try_from).collect::<Result<Vec<_>, _>>()?;
```
Just a suggestion

File: mirrord/cli/src/execution.rs [hunk 2]
```diff
@@ -164,6 +169,25 @@ impl MirrordExecution {
...
+                    ..
+                }) => Some(version.clone()),
+                AgentConnectInfo::DirectKubernetes(_) => {
+                    Some(MirrordExecution::get_agent_version(&mut connection).await?)
+                }
+                _ => None,
+            };
+            if version.is_some() && !HTTP_COMPOSITE_FILTER_VERSION.matches(&version.unwrap()) {
+                Err(ConfigError::Conflict(format!(
+                    "Cannot use 'any' or 'all' incoming filter types fields with versions of mirrord agent lower than {}",
```
Comment (Razz4780) [inline mirrord/cli/src/execution.rs:185]:
```suggestion
                    "Cannot use 'any_of' or 'all_of' HTTP filter types, protocol version used by mirrord-agent must match {}. Consider using a newer version of mirrord-agent",
```

File: mirrord/cli/src/execution.rs [hunk 3]
```diff
@@ -164,6 +169,25 @@ impl MirrordExecution {
...
+                AgentConnectInfo::Operator(OperatorSession {
+                    operator_protocol_version: Some(version),
+                    ..
+                }) => Some(version.clone()),
+                AgentConnectInfo::DirectKubernetes(_) => {
+                    Some(MirrordExecution::get_agent_version(&mut connection).await?)
+                }
+                _ => None,
+            };
+            if version.is_some() && !HTTP_COMPOSITE_FILTER_VERSION.matches(&version.unwrap()) {
```
Comment (Razz4780) [inline mirrord/cli/src/execution.rs:183]:
```suggestion
            if !version.map(|version| HTTP_COMPOSITE_FILTER_VERSION.matches(&version)).unwrap_or(false) {
```

File: mirrord/cli/src/execution.rs [hunk 4]
```diff
@@ -275,6 +299,28 @@ impl MirrordExecution {
...
+    async fn get_agent_version(connection: &mut AgentConnection) -> Result<Version> {
+        let Ok(_) = connection
+            .sender
+            .send(ClientMessage::SwitchProtocolVersion(
+                mirrord_protocol::VERSION.clone(),
+            ))
+            .await
+        else {
+            return Err(CliError::InitialAgentCommFailed(
+                "protocol version check failed".to_string(),
```
Comment (Razz4780) [inline mirrord/cli/src/execution.rs:311]:
```suggestion
                "failed to send protocol version request".to_string(),
```

File: mirrord/cli/src/execution.rs [hunk 5]
```diff
@@ -275,6 +299,28 @@ impl MirrordExecution {
...
+            ));
+        };
+        if let Some(DaemonMessage::SwitchProtocolVersionResponse(version)) =
+            connection.receiver.recv().await
+        {
+            return Ok(version);
+        }
+        return Err(CliError::InitialAgentCommFailed(
+            "protocol version check failed".to_string(),
+        ));
```
Comment (Razz4780) [inline mirrord/cli/src/execution.rs:321]: This should be a `match` to provide better errors (like in`fetch_env_vars`)

File: mirrord/config/src/feature/network/incoming/http_filter.rs [hunk 6]
```diff
@@ -124,6 +165,10 @@ impl MirrordToggleableConfig for HttpFilterFileConfig {
             .source_value(context)
             .transpose()?;
 
+        // TODO: make sure env vars should be used to set these
+        let all_of = None;
+        let any_of = None;
```
Comment (Razz4780) [inline mirrord/config/src/feature/network/incoming/http_filter.rs:170]: Can be removed, I think

**Review by Razz4780 — CHANGES_REQUESTED** [review]
Only minor stuff

You can regenerate schema with
```
cargo test -p mirrord-config check_schema_file_exists_and_is_valid_or_create_it -- --ignored --nocapture
```
You can also regenerate `configuration.md` with
```
cargo run --bin medschool -- -i ./mirrord/config/src -o ./mirrord/config/configuration.md 
```

File: mirrord/agent/src/steal/http/filter.rs (hunk 1 above)
Comment (gememma) [reply mirrord/agent/src/steal/http/filter.rs:37]: i spent so long trying to get map to work 😆 this is much nicer

**Pull request merged**